Invalid replacement list -> async generator

Pyupgrade applied this change for me:

```diff
diff --git a/tests/unit/test_repair.py b/tests/unit/test_repair.py
index 3a153bc..d1658e5 100644
--- a/tests/unit/test_repair.py
+++ b/tests/unit/test_repair.py
@@ -24,15 +24,15 @@ async def test_repair_uids(uid):
         "two": ("asdf", Item(f"BEGIN:VCARD\nFN:Peppi\nUID:{uid}\nEND:VCARD")),
     }

-    uid1, uid2 = [(await s.get(href))[0].uid async for href, etag in s.list()]
+    uid1, uid2 = ((await s.get(href))[0].uid async for href, etag in s.list())
     assert uid1 == uid2

     await repair_storage(s, repair_unsafe_uid=False)

-    uid1, uid2 = [
+    uid1, uid2 = (
         (await s.get(href))[0].uid
         for href, etag in await aiostream.stream.list(s.list())
-    ]
+    )
     assert uid1 != uid2
```

This run in its original form (the list is expanded), but doesn't on the latter:

```
tests/unit/test_repair.py:27: in test_repair_uids
    uid1, uid2 = ((await s.get(href))[0].uid async for href, etag in s.list())
E   TypeError: cannot unpack non-iterable async_generator object
```

I'm using `pyupgrade 2.23.2`, with `--py37-plus`.

--- a/pyupgrade/_ast_helpers.py
+++ b/pyupgrade/_ast_helpers.py
@@ -50,3 +50,10 @@ def contains_await(node: ast.AST) -> bool:
             return True
     else:
         return False
+
+
+def is_async_listcomp(node: ast.ListComp) -> bool:
+    return (
+        any(gen.is_async for gen in node.generators) or
+        contains_await(node)
+    )

--- a/pyupgrade/_plugins/generator_expressions_pep289.py
+++ b/pyupgrade/_plugins/generator_expressions_pep289.py
@@ -8,7 +8,7 @@ from tokenize_rt import Offset
 from tokenize_rt import Token
 
 from pyupgrade._ast_helpers import ast_to_offset
-from pyupgrade._ast_helpers import contains_await
+from pyupgrade._ast_helpers import is_async_listcomp
 from pyupgrade._data import register
 from pyupgrade._data import State
 from pyupgrade._data import TokenFunc
@@ -53,11 +53,7 @@ def visit_Call(
             node.func.id in ALLOWED_FUNCS and
             node.args and
             isinstance(node.args[0], ast.ListComp) and
-            not any(
-                generator.is_async
-                for generator in node.args[0].generators
-            ) and
-            not contains_await(node.args[0])
+            not is_async_listcomp(node.args[0])
     ):
         if len(node.args) == 1 and not node.keywords:
             yield ast_to_offset(node.args[0]), _delete_list_comp_brackets

--- a/pyupgrade/_plugins/unpack_list_comprehension.py
+++ b/pyupgrade/_plugins/unpack_list_comprehension.py
@@ -7,6 +7,7 @@ from tokenize_rt import Offset
 from tokenize_rt import Token
 
 from pyupgrade._ast_helpers import ast_to_offset
+from pyupgrade._ast_helpers import is_async_listcomp
 from pyupgrade._data import register
 from pyupgrade._data import State
 from pyupgrade._data import TokenFunc
@@ -31,6 +32,7 @@ def visit_Assign(
             state.settings.min_version >= (3,) and
             len(node.targets) == 1 and
             isinstance(node.targets[0], ast.Tuple) and
-            isinstance(node.value, ast.ListComp)
+            isinstance(node.value, ast.ListComp) and
+            not is_async_listcomp(node.value)
     ):
         yield ast_to_offset(node.value), _replace_list_comprehension

--- a/pyupgrade/_plugins/unpacking_argument_list_comprehensions.py
+++ b/pyupgrade/_plugins/unpacking_argument_list_comprehensions.py
@@ -5,6 +5,7 @@ from typing import Tuple
 from tokenize_rt import Offset
 
 from pyupgrade._ast_helpers import ast_to_offset
+from pyupgrade._ast_helpers import is_async_listcomp
 from pyupgrade._data import register
 from pyupgrade._data import State
 from pyupgrade._data import TokenFunc
@@ -19,6 +20,7 @@ def visit_Starred(
 ) -> Iterable[Tuple[Offset, TokenFunc]]:
     if (
         state.settings.min_version >= (3,) and
-        isinstance(node.value, ast.ListComp)
+        isinstance(node.value, ast.ListComp) and
+        not is_async_listcomp(node.value)
     ):
         yield ast_to_offset(node.value), replace_list_comp_brackets

--- a/tests/features/unpack_list_comprehension_test.py
+++ b/tests/features/unpack_list_comprehension_test.py
@@ -17,6 +17,11 @@ from pyupgrade._main import _fix_plugins
             (3,),
             id='assignment to single variable',
         ),
+        pytest.param(
+            'x, = [await foo for foo in bar]',
+            (3,),
+            id='async comprehension',
+        ),
     ),
 )
 def test_fix_typing_text_noop(s, version):

--- a/tests/features/unpacking_argument_list_comprehensions_test.py
+++ b/tests/features/unpacking_argument_list_comprehensions_test.py
@@ -32,6 +32,11 @@ from pyupgrade._main import _fix_plugins
             (3,),
             id='Starred, no list comp',
         ),
+        pytest.param(
+            'foo(*[x async for x in bar])',
+            (3,),
+            id='async listcomp',
+        ),
     ),
 )
 def test_fix_unpack_argument_list_comp_noop(s, version):

